src/main/java/org/opensearch/knn/index/codec/util/KNNCodecUtil.java

```diff
@@ -84,4 +92,67 @@ public static List<String> getEngineFiles(String extension, String fieldName, Se
             .collect(Collectors.toList());
         return engineFiles;
     }
+
+    /**
+     * Get engine file name from given field and segment info.
+     * Ex: _0_165_my_field.faiss
+     *
+     * @param field : Field info that might have a vector index file. Not always it has it.
+     * @param segmentInfo : Segment where we are collecting an engine file list.
+     * @return : Found vector engine names, if not found, returns null.
+     */
+    public static String getNativeEngineFileFromFieldInfo(FieldInfo field, SegmentInfo segmentInfo) {
+        if (!field.attributes().containsKey(KNN_FIELD)) {
+            return null;
+        }
+        // Only Native Engine put into indexPathMap
+        final KNNEngine knnEngine = getNativeKNNEngine(field);
+        if (knnEngine == null) {
+            return null;
+        }
+        final List<String> engineFiles = KNNCodecUtil.getEngineFiles(knnEngine.getExtension(), field.name, segmentInfo);
+        if (engineFiles.isEmpty()) {
+            return null;
+        } else {
+            final String vectorIndexFileName = engineFiles.get(0);
+            return vectorIndexFileName;
+        }
+    }
+
+    public static List<String> getVectorCacheKeysFromSegmentReaderState(SegmentReadState segmentReadState) {
+        final List<String> cacheKeys = new ArrayList<>();
+
+        for (FieldInfo field : segmentReadState.fieldInfos) {
+            // Only segments that contains BinaryDocValues and doesn't have vector values should be considered.
+            // By default, we don't create BinaryDocValues for knn field anymore. However, users can set doc_values = true
+            // to create binary doc values explicitly like any other field. Hence, we only want to include fields
+            // where approximate search is possible only by BinaryDocValues.
+            if (field.getDocValuesType() != DocValuesType.BINARY || field.hasVectorValues()) {
```

Comment:
does this still work for `NativeEngine990KNNVectorReader`? Considering [this](https://github.com/opensearch-project/k-NN/pull/2182/files/88509d4a71ee215ad9dd88473a5b2a36da2e7cb8#diff-e8c0b3983b0fd2edc56cabf9fbb439ea56eaa8917e7bff5713a30ee5e2a4e69cR64) shouldn't this have a `!field.hasVectorValues`? Or did I misunderstand something here?

Reply:
no it won't work for 990KNNVectorReader
will update the fix shortly.

Reply:
I think it still needs to be done in constructor.
To extract KNNEngine info, it access cluster state which can be not available for some cases apparently.
Will revert it back to have it in constructor.

```
Suppressed: java.lang.AssertionError: should not be called by a cluster state applier. reason [the applied cluster state is not yet available]
»               at org.opensearch.cluster.service.ClusterApplierService.assertNotCalledFromClusterStateApplier(ClusterApplierService.java:446) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»               at org.opensearch.cluster.service.ClusterApplierService.state(ClusterApplierService.java:230) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»               at org.opensearch.cluster.service.ClusterService.state(ClusterService.java:183) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»               at org.opensearch.knn.indices.ModelDao$OpenSearchKNNModelDao.getMetadata(ModelDao.java:457) ~[?:?]
»               at org.opensearch.knn.indices.ModelUtil.getModelMetadata(ModelUtil.java:52) ~[?:?]
»               at org.opensearch.knn.common.FieldInfoExtractor.extractKNNEngine(FieldInfoExtractor.java:41) ~[?:?]
»               at org.opensearch.knn.index.codec.util.KNNCodecUtil.getNativeKNNEngine(KNNCodecUtil.java:126) ~[?:?]
»               at org.opensearch.knn.index.codec.util.KNNCodecUtil.getNativeEngineFileFromFieldInfo(KNNCodecUtil.java:106) ~[?:?]

```